README.md
# redis-resource-wait-list

[![NPM version][npm-image]][npm-url]
[![Build Status][travis-image]][travis-url]
[![Coveralls Status][coveralls-image]][coveralls-url]
[![Dependency Status][depstat-image]][depstat-url]
[![Downloads][download-badge]][npm-url]

> Manage limited atomic resource using redis, with wait / queue feature.

## Install

```sh
npm i -S redis-resource-wait-list
```

## Usage

```js
const List = require('redis-resource-wait-list');

// options and defaults
const list = List({
  name: 'server-list',
  resources: ['server-1', 'server-2', 'server-3'],
  options: {
    maxTimeoutToRelease: 5 * 60 * 1000, // Defaults to five minutes, important. see Reliable queues.
    maxTimeoutToWait: 5 * 60 * 1000, // Defaults to five minutes, important. see Reliable queues.
    intervalToCheckRelease: 30000, // Resource that exceed maxTimeoutToRelease will release every check, thus the interval option
    redisUrl: 'redis://127.0.0.1:6379',
    redisOptions: {},
    redisPrefix: 'wl',
  },
});
list.start() //Promise
  .then(() => list.acquire())
  .then((resource) => {
    // use resource
    return list.release(resource); // Promise
    list.add('server-4') // Promise
      .then(() => console.log('added'));
    list.remove('server-1') // Promise
      .then(() => console.log('removed'));
    list.getInfo()
      .then((info) => console.log('info'));
    /*
    info = {
      resources: ['server-2', 'server-3', 'server-4'],
      available: ['server-2', 'server-3', 'server-4'],
      busy: [],
      // clientWaiting: 0, // not implemented
      settings,
    };
    */
    list.stop(); // Promise

  });
```

## License

MIT © [Raabb Ajam](https://github.com/raabbajam)

[npm-url]: https://npmjs.org/package/redis-resource-wait-list
[npm-image]: https://img.shields.io/npm/v/redis-resource-wait-list.svg?style=flat-square

[travis-url]: https://travis-ci.org/raabbajam/redis-resource-wait-list
[travis-image]: https://img.shields.io/travis/raabbajam/redis-resource-wait-list.svg?style=flat-square

[coveralls-url]: https://coveralls.io/r/raabbajam/redis-resource-wait-list
[coveralls-image]: https://img.shields.io/coveralls/raabbajam/redis-resource-wait-list.svg?style=flat-square

[depstat-url]: https://david-dm.org/raabbajam/redis-resource-wait-list
[depstat-image]: https://david-dm.org/raabbajam/redis-resource-wait-list.svg?style=flat-square

[download-badge]: http://img.shields.io/npm/dm/redis-resource-wait-list.svg?style=flat-square


CHANGELOG.md
<!-- START doctoc generated TOC please keep comment here to allow auto update -->
<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->
# Changelog

- [v1.0.0](#v100)

<!-- END doctoc generated TOC please keep comment here to allow auto update -->

### v1.0.0

 * redis-resource-wait-list initial commit.


test/index.js
/* eslint no-console: 0 */
const Promise = require('bluebird');
const test = require('tape');
const List = require('../src');
test('List -- start', (assert) => {
  const list = List({
    name: 'resource-x',
    resources: ['x1'],
  });
  return list.start()
    .then(() => list.getInfo()
      .then((info) => {
        assert.ok(info.available.length === 1, 'should make available 1');
        assert.ok(info.busy.length === 0, 'should make busy length 0');
        return list.destroy().then(() => list.stop()).then(() => assert.end());
      })
    )
    .catch((error) => {
      list.destroy().then(() => list.stop())
        .then(() => assert.end(error));
    });
});
test('List -- acquire', (assert) => {
  const list = List({
    name: 'resource-x',
    resources: ['x1'],
  });
  return list.start()
    .then(() => list.acquire())
    .then((x1) => {
      assert.ok(x1 === 'x1');
      return list.getInfo()
        .then((info) => {
          assert.ok(info.available.length === 0, 'should make none available');
          assert.ok(info.busy[0] === 'x1', 'should register resource as busy');
          assert.ok(info.busy.length === 1, 'should make busy length 1');
          return list.destroy().then(() => list.stop()).then(() => assert.end());
        });
    })
    .catch((error) => {
      list.destroy().then(() => list.stop())
        .then(() => assert.end(error));
    });
});
test('List -- release', (assert) => {
  const list = List({
    name: 'resource-x',
    resources: ['x1'],
  });
  return list.start()
    .then(() => list.acquire())
    .then((x1) => list.release(x1))
    .then(() => list.getInfo())
    .then((info) => {
      assert.ok(info.available.length === 1, 'should make resource available');
      assert.ok(info.busy.length === 0, 'should release busy resource');
      return list.destroy().then(() => list.stop()).then(() => assert.end());
    })
    .catch((error) => {
      list.destroy().then(() => list.stop())
        .then(() => assert.end(error));
    });
});
test('List -- add', (assert) => {
  const list = List({
    name: 'resource-x',
    resources: ['x1'],
  });
  return list.start()
    .then(() => list.add('x2'))
    .then(() => list.getInfo())
    .then((info) => {
      assert.ok(info.available.length === 2, 'should make available 2');
      return list.destroy().then(() => list.stop()).then(() => assert.end());
    })
    .catch((error) => {
      list.destroy().then(() => list.stop())
        .then(() => assert.end(error));
    });
});
test('List -- add multiple', (assert) => {
  const list = List({
    name: 'resource-x',
    resources: ['x1'],
  });
  return list.start()
    .then(() => list.add('x2', 'x3'))
    .then(() => list.getInfo())
    .then((info) => {
      assert.ok(info.available.length === 3, 'should make available 3');
      return list.destroy().then(() => list.stop()).then(() => assert.end());
    })
    .catch((error) => {
      list.destroy().then(() => list.stop())
        .then(() => assert.end(error));
    });
});
test('List -- remove', (assert) => {
  const list = List({
    name: 'resource-x',
    resources: ['x1'],
  });
  return list.start()
    .then(() => list.remove('x1'))
    .then(() => list.getInfo())
    .then((info) => {
      assert.ok(info.available.length === 0, 'should make available 0');
      return list.destroy().then(() => list.stop()).then(() => assert.end());
    })
    .catch((error) => {
      list.destroy().then(() => list.stop())
        .then(() => assert.end(error));
    });
});
test('List -- remove multiple', (assert) => {
  const list = List({
    name: 'resource-x',
    resources: ['x1', 'x2'],
  });
  return list.start()
    .then(() => list.remove('x1', 'x2'))
    .then(() => list.getInfo())
    .then((info) => {
      assert.ok(info.available.length === 0, 'should make available 0');
      return list.destroy().then(() => list.stop()).then(() => assert.end());
    })
    .catch((error) => {
      list.destroy().then(() => list.stop())
        .then(() => assert.end(error));
    });
});
test('List -- stop', (assert) => {
  const list = List({
    name: 'resource-x',
    resources: ['x1'],
  });
  return list.start()
    .then(() => list.destroy().then(() => list.stop()))
    .then(() => list.getInfo())
    .catch((error) => {
      assert.ok(Boolean(error), 'should error connection closed');
      assert.end();
    });
});
test('List -- started in 2 places', (assert) => {
  const list = List({
    name: 'resource-x',
    resources: ['x1'],
  });
  return list.start()
    .then(() => list.acquire())
    .then(() => list.start())
    .then(() => list.getInfo())
    .then((info) => {
      assert.ok(info.available.length === 0, 'should not re init');
      return list.destroy().then(() => list.stop()).then(() => assert.end());
    })
    .catch((error) => {
      list.destroy().then(() => list.stop())
        .then(() => assert.end(error));
    });
});
test('List -- acquire more than available', (assert) => {
  const delay = 100;
  const list = List({
    name: 'resource-x',
    resources: ['x1'],
  });
  return list.start()
    .then(() => list.acquire())
    .then((x1) => {
      Promise.delay(delay).then(() => list.release(x1));
      return list.acquire();
    })
    .then((x1) => {
      assert.ok(x1 === 'x1', 'should wait until available');
      return list.destroy().then(() => list.stop()).then(() => assert.end());
    })
    .catch((error) => {
      list.destroy().then(() => list.stop())
        .then(() => assert.end(error));
    });
});
test('List -- maxTimeoutToRelease', (assert) => {
  const delay = 2000;
  const list = List({
    name: 'resource-x',
    resources: ['x1'],
    options: {
      maxTimeoutToRelease: 1000,
      intervalToCheckRelease: 1000,
    },
  });
  return list.start()
    .then(() => list.acquire())
    .delay(delay)
    .then(() => list.getInfo())
    .then((info) => {
      assert.ok(info.available.length === 1, 'should released due to timeout');
      return list.destroy().then(() => list.stop()).then(() => assert.end());
    })
    .catch((error) => {
      list.destroy().then(() => list.stop())
        .then(() => assert.end(error));
    });
});
test('List -- removed when busy', (assert) => {
  const list = List({
    name: 'resource-x',
    resources: ['x1'],
  });
  return list.start()
    .then(() => list.acquire())
    .then((x1) => list.remove(x1).then(() => list.release(x1)))
    .then(() => list.getInfo())
    .then((info) => {
      assert.ok(info.available.length === 0, 'should not released back to available due to removed');
      return list.destroy().then(() => list.stop()).then(() => assert.end());
    })
    .catch((error) => {
      list.destroy().then(() => list.stop())
        .then(() => assert.end(error));
    });
});
test('List -- released multiple times', (assert) => {
  const list = List({
    name: 'resource-x',
    resources: ['x1'],
  });
  return list.start()
    .then(() => list.acquire())
    .then((x1) => list.release(x1).then(() => list.release(x1)).then(() => list.release(x1)))
    .then(() => list.getInfo())
    .then((info) => {
      assert.ok(info.available.length === 1, 'should not duplicate on available');
      return list.destroy().then(() => list.stop()).then(() => assert.end());
    })
    .catch((error) => {
      list.destroy().then(() => list.stop())
        .then(() => assert.end(error));
    });
});
test('List -- sync', (assert) => {
  const list = List({
    name: 'resource-x',
    resources: ['x1', 'x2'],
  });
  return list.start()
    .then(() => list.sync(['x1', 'x3']))
    .then(() => list.getInfo())
    .then((info) => {
      assert.ok(info.available.length === 2, 'should sync');
      return list.destroy().then(() => list.stop()).then(() => assert.end());
    })
    .catch((error) => {
      list.destroy().then(() => list.stop())
        .then(() => assert.end(error));
    });
});


src/index.js
const debug = require('debug')('redis-resource-wait-list:main');
const Promise = require('bluebird');
const moment = require('moment');
const pool = require('generic-promise-pool');
const redis = require('redis');
const _ = require('lodash');
const bigNumber = 9999;
Promise.promisifyAll(redis.RedisClient.prototype);
Promise.promisifyAll(redis.Multi.prototype);
function List(data = {}) {
  const {
    name,
    resources = [],
    options = {},
  } = data;
  const defaults = {
    maxTimeoutToRelease: 300000,
    maxTimeoutToWait: 300000,
    intervalToCheckRelease: 30000,
    redisUrl: 'redis://127.0.0.1:6379',
    redisOptions: {},
    redisPrefix: 'wl',
  };
  const settings = Object.assign({}, defaults, options);
  const {redisUrl, redisOptions, redisPrefix} = settings;
  const poolOptions = {
    name: `${name}-pool`,
    create: () => Promise.resolve(redis.createClient(redisUrl, redisOptions)),
    destroy: (redisClient) => redisClient.quitAsync(),
    max: 30,
    min: 1,
  };
  const redisPool = pool.create(poolOptions);
  const resourcesListKey = `${redisPrefix}:${name}:resource-set`;
  const availableListKey = `${redisPrefix}:${name}:available-list`;
  const busyListKey = `${redisPrefix}:${name}:busy-list`;
  const busySetKey = `${redisPrefix}:${name}:busy-set`;
  const keys = {
    resourcesListKey,
    availableListKey,
    busyListKey,
    busySetKey,
  };
  const properties = {
    stopped: false,
    redisPool,
    settings,
    name,
    resources,
    keys,
  };
  const prototypes = {
    getRedisClient,
    start,
    acquire,
    release,
    add,
    remove,
    getInfo,
    releaseAllExpired,
    stop,
    destroy,
    sync,
  };
  const list = Object.assign(Object.create(prototypes), properties);
  return list;
}

module.exports = List;

function start() {
  debug('start');
  const {keys, resources = [], settings} = this;
  const {intervalToCheckRelease} = settings;
  return this.getRedisClient((redisClient) =>
    redisClient.existsAsync(keys.resourcesListKey)
      .then((exist) => {
        if (exist) {
          return this.sync(resources);
        }
        const repeat = () => this.releaseAllExpired()
          .then(() => {
            this.timer = setTimeout(repeat, intervalToCheckRelease);
          })
          .catch(() => {});
        this.timer = setTimeout(repeat, intervalToCheckRelease);
        if (!resources.length) {
          return null;
        }
        return redisClient.multi()
          .sadd(keys.resourcesListKey, ...resources)
          .rpush(keys.availableListKey, ...resources)
          .execAsync();
      }))
      .return(this);
}

function acquire() {
  debug('acquire');
  const {keys, settings} = this;
  const {maxTimeoutToWait, maxTimeoutToRelease} = settings;
  const thousand = 1000;
  const timeout = Math.ceil(maxTimeoutToWait / thousand);
  return this.getRedisClient((redisClient) =>
    redisClient.scardAsync(keys.resourcesListKey)
      .then((hasResource) => {
        if (!hasResource) {
          const error = new Error('ENOTFOUND. This list has no member');
          throw error;
        }
        return redisClient.brpoplpushAsync(keys.availableListKey, keys.busyListKey, timeout);
      })
      .tap((resource) => {
        if (!resource) {
          const error = new Error('ETIMEOUT. Can\'t find resouce.');
          throw error;
        }
        const expired = moment().add(maxTimeoutToRelease, 'ms').unix();
        redisClient.zaddAsync(keys.busySetKey, expired, resource);
      }));
}

function release(resource) {
  debug(`release ${resource}`);
  const {keys} = this;
  const sha = '7834e4d113867947e1e2e376f9623297bfab7991';
  const restArgs = [
    4,
    keys.busyListKey, keys.busySetKey, keys.resourcesListKey, keys.availableListKey,
    resource,
  ];
  return this.getRedisClient((redisClient) =>
    redisClient.evalshaAsync(sha, ...restArgs)
      .catch((error) => {
        if (!/NOSCRIPT/i.test(error.message)) {
          throw error;
        }
        // eslint-disable-next-line max-len
        const source = 'redis.call("LREM", KEYS[1], 9999, ARGV[1]) redis.call("ZREM", KEYS[2], ARGV[1]) if redis.call("SISMEMBER", KEYS[3], ARGV[1]) == 1 then redis.call("LREM", KEYS[4], 9999, ARGV[1]) return redis.call("RPUSH", KEYS[4], ARGV[1]) else return nil end';
        return redisClient.evalAsync(source, ...restArgs);
      }));
}

function add(...resources) {
  debug(`add ${resources}`);
  const {keys} = this;
  return this.getRedisClient((redisClient) => {
    const multi = redisClient.multi();
    resources.forEach((resource) => {
      multi
        .sadd(keys.resourcesListKey, resource)
        .lrem(keys.availableListKey, bigNumber, resource)
        .rpush(keys.availableListKey, resource);
    });
    return multi.execAsync();
  });
}

function remove(...resources) {
  debug(`remove ${resources}`);
  const {keys} = this;
  return this.getRedisClient((redisClient) => {
    const multi = redisClient.multi();
    resources.forEach((resource) => {
      multi
        .lrem(keys.busyListKey, bigNumber, resource)
        .lrem(keys.availableListKey, bigNumber, resource)
        .srem(keys.resourcesListKey, resource);
    });
    return multi.execAsync();
  });
}

function getInfo() {
  debug('getInfo');
  const {keys, settings, name} = this;
  return this.getRedisClient((redisClient) =>
    redisClient.multi()
      .smembers(keys.resourcesListKey)
      .lrange(keys.availableListKey, 0, bigNumber)
      .lrange(keys.busyListKey, 0, bigNumber)
      .execAsync())
    .spread((resources, available, busy) => ({name, resources, available, busy, settings}));
}

function stop() {
  debug('stop');
  const {redisPool, timer} = this;
  if (timer) {
    clearTimeout(timer);
  }
  return redisPool.drain()
    .then(() => {
      this.stopped = true;
    });
}

function destroy(options = {}) {
  debug('destroy');
  const {keys} = this;
  return this.getRedisClient((redisClient) =>
    redisClient.multi()
      .del(keys.resourcesListKey)
      .del(keys.availableListKey)
      .del(keys.busyListKey)
      .del(keys.busySetKey)
      .execAsync())
    .tap(() => {
      const {with_stop} = options;
      if (with_stop) {
        return this.stop();
      }
      return null;
    });
}

function getRedisClient(fn) {
  const {redisPool, stopped} = this;
  return Promise.resolve()
    .then(() => {
      if (stopped) {
        const error = new Error('The resource list is already stopped and inaccessible.');
        throw error;
      }
    })
    .then(() => redisPool.acquire(fn));
}

function releaseAllExpired() {
  debug('releaseAllExpired');
  const {keys, stopped} = this;
  return Promise.resolve()
    .then(() => {
      if (stopped) {
        const error = new Error('The resource list is already stopped and inaccessible.');
        throw error;
      }
      return this.getRedisClient((redisClient) => {
        const now = moment().unix();
        return redisClient.zrangebyscoreAsync(keys.busySetKey, 0, now)
          // .tap(debug)
          .map((resource) => this.release(resource));
      });
    });
}

function sync(resoucesReference) {
  debug('sync');
  return this.getInfo()
    .then((info) => {
      const {resources} = info;
      const resourcesToRemove = _.difference(resources, resoucesReference);
      const resourcesToAdd = _.difference(resoucesReference, resources);
      return Promise.all([
        this.remove(...resourcesToRemove),
        this.add(...resourcesToAdd),
      ]);
    });
}
